state-chain/pallets/cf-elections/src/electoral_systems/mocks.rs

```diff
@@ -0,0 +1,321 @@
+	}
+
+	pub fn access(&self) -> &MockAccess<ES> {
+		&self.electoral_access
+	}
+
+	#[track_caller]
+	fn inner_force_consensus_update(
+		self,
+		election_id: ElectionIdentifierOf<ES>,
+		new_consensus: ConsensusStatus<ES::Consensus>,
+	) -> Self {
+		let mut electoral_access = self.electoral_access.clone();
+		electoral_access
+			.election_mut(election_id)
+			.unwrap()
+			.set_consensus_status(new_consensus);
+
+		Self { electoral_access, ..self }
+	}
+
+	/// Test the finalization of the election.
+	///
+	/// `pre_finalize_checks` is a closure that is called with a read-only access to the electoral
+	/// state before finalization.
+	///
+	/// `post_finalize_checks` is a list of checks that are run after finalization. These checks are
+	///
+	/// See [register_checks] and
+	#[track_caller]
+	pub fn test_on_finalize(
+		mut self,
+		on_finalize_context: &ES::OnFinalizeContext,
+		pre_finalize_checks: impl FnOnce(&MockAccess<ES>),
+		post_finalize_checks: impl IntoIterator<Item = Check<ES>>,
+	) -> Self {
+		let pre_finalize = self.electoral_access.clone();
+		// TODO: Move 'hook' static local checks into MockAccess so we can remove this.
+		pre_finalize_checks(&pre_finalize);
+
+		self.electoral_access.finalize_elections(on_finalize_context).unwrap();
+
+		let post_finalize = self.electoral_access.clone();
+		for check in post_finalize_checks {
+			check.check(&pre_finalize, &post_finalize);
+		}
+		self
+	}
+}
+
+/// Allows registering checks for an electoral system. Once registered, the checks can be used
+/// through the `Check` struct.
+///
+/// Example:
+///
+/// ```ignore
+/// register_checks! {
+///     MonotonicMedianTest {
+///         monotonically_increasing_state(pre_finalize, post_finalize) {
+///             assert!(
+///                 post_finalize.unsynchronised_state().unwrap() >= pre_finalize.unsynchronised_state().unwrap(),
+///                 "Expected state to increase post-finalization."
+///             );
+///         },
+///         // ..
+///     }
+/// }
+/// ```
+///
+///
+/// Alternatively, you can specify extra constraints for the electoral system instead of using a
+/// concrete type:
+///
+/// ```ignore
+/// register_checks! {
+///     #[ extra_constraints: ES: ElectoralSystem, ES::ElectionIdentifierExtra: Default ]#
+///     monotonically_increasing_state(pre_finalize, post_finalize) {
+///         assert!(
+///             post_finalize.unsynchronised_state().unwrap() >= pre_finalize.unsynchronised_state().unwrap(),
+///             "Expected state to increase post-finalization."
+///         );
+///     },
+/// }
+/// ```
+#[macro_export]
+macro_rules! register_checks {
+	(
+		$system:ident {
+			$(
+				$check_name:ident($arg_1:ident, $arg_2:ident) $check_body:block
+			),+ $(,)*
+		}
+	) => {
+		impl Check<$system>{
+			$(
+				pub fn $check_name() -> Self {
+					Self::new(#[track_caller] |$arg_1, $arg_2| $check_body)
+				}
+			)+
+		}
+	};
+	(
+		$(
+			#[ extra_constraints: $( $t:ty : $tc:path ),+ ]#
```

Comment:
nit:
(and in the comment above too)
```suggestion
			#[extra_constraints: $( $t:ty : $tc:path ),+]#
```